Fix AsyncStream leak in kubernetes watch (#1694)

sgrankin opened the pull request:
k8s watch returned an AsyncStream, the head of which was leaking (as
determined via heap dumps), leading to significant leaks over 24-48hrs.

Since the AsyncStream was not strictly necessary, rewrite `watch` as
an update function suitable for Var.async (as used by `activity`).
- The change attempts to preserve previous bug fixes and comments
  documenting them.
- The AsyncStream enrichments and their tests have been removed as they
  are no longer used.
- A `watch` helper was added to ApiTest to hide the more complicated
  usage pattern around .watch.
- Fix orphaned infinite-retry watches in ingress tests by adding
  no-data /watch/ handlers.

Fixes #1694

siggy commented:
Ran this through our integration test suite. Results look good, memory is stable:
<img width="1267" alt="screen shot 2017-11-27 at 12 11 49 pm" src="https://user-images.githubusercontent.com/236915/33288012-8fea4cda-d36f-11e7-81b4-70a97be3475e.png">

olix0r commented on k8s/src/main/scala/io/buoyant/k8s/Watchable.scala:
```diff
@@ -83,19 +83,19 @@ private[k8s] abstract class Watchable[O <: KubeObject: TypeReference, W <: Watch
-   *         the watch to be terminated.
+   * @return An update function that may be passed to [[Var.async]] to create stream of [[Watch]] objects containing
+   *         additions, modifications, and deletions to the items in the list being watched, and a Closable handle
+   *         allowing for the watch to be terminated.
```
nit: i suspect these comment lines are a big longer than the usual 100 char limit we like to apply.

Would you mind wrapping comments at 100 chars here & throughout?  Thanks!

olix0r commented on k8s/src/main/scala/io/buoyant/k8s/Watchable.scala:
```diff
@@ -83,19 +83,19 @@ private[k8s] abstract class Watchable[O <: KubeObject: TypeReference, W <: Watch
     fieldSelector: Option[String] = None,
     resourceVersion: Option[String] = None
-  ): (AsyncStream[W], Closable) = {
+  ): (Activity.State[W] => Unit) => Closable = { (output: Activity.State[W] => Unit) =>
```
nit: probably most idiomatic to call _output_ `state`

siggy commented:
memory usage flat after 24 hours. that drop at the end is me forcing a gc with `jmap`:
![screen shot 2017-11-28 at 10 46 03 am](https://user-images.githubusercontent.com/236915/33337967-7f974224-d429-11e7-8d75-712c5714e2bd.png)

hawkw commented on k8s/src/main/scala/io/buoyant/k8s/Watchable.scala:
```diff
@@ -139,71 +121,108 @@ private[k8s] abstract class Watchable[O <: KubeObject: TypeReference, W <: Watch
-        .increasingOnly
+    }
+
+    def _processEventStream(
```
A comment summarizing this function's behaviour might be nice.

hawkw commented on k8s/src/main/scala/io/buoyant/k8s/Watchable.scala:
```diff
@@ -139,71 +121,108 @@ private[k8s] abstract class Watchable[O <: KubeObject: TypeReference, W <: Watch
+      previousEvent: Option[W] = None
+    ): Future[Unit] = {
+      stream().uncons.flatMap {
+        case Some((w, ws)) =>
```
Take it or leave it, but I think this _might_ be a little clearer if we collapsed the nested `match` expression into one `match` with guards, like:
```scala
// special case to handle Kubernetes bug where "too old
// resource version" errors are returned with status code 200
// rather than status code 410.
// see https://github.com/kubernetes/kubernetes/issues/35068
// for details
case Some((e: Watch.Error[0], ws))  if e.status.code.contains(410) =>
  log.debug(
   "k8s returned 'too old resource version' error with " +
     "incorrect HTTP status code, restarting watch"
  )
  _resourceVersionTooOld()
  
case Some((event, ws)) =>
  import Ordering.Implicits._
  if (previousEvent.forall((_: W) < event)) {
    state(Activity.Ok(event))
    _processEventStream(ws, event.resourceVersion, Some(event))
  } else {
    // Ignore any events where we receive a resource version lower
    // than the last received resource version.
    _processEventStream(ws, resourceVersion, previousEvent)
  }

// if the stream ends (k8s will kill connections after ~30m), restart it.
case None if resourceVersion.isDefined => _watch(resourceVersion)

// In this case, we want to try loading the initial information instead before
// watching again.
case None => _resourceVersionTooOld()
```

sgrankin replied:
(Also rephrased the comments slightly for clarity)

hawkw commented on k8s/src/test/scala/io/buoyant/k8s/WatchableTest.scala:
```diff
@@ -1,24 +0,0 @@
-import io.buoyant.test.FunSuite
-import org.scalatest.prop.GeneratorDrivenPropertyChecks
-
-class WatchableTest extends FunSuite
```
Since `IncreasingOnlyAsyncStream` was removed, I'd like to see another test somewhere that asserts that the events on `Watchable.Activity` have monotonically-increasing resource versions even if the Kubernetes API sends events out of order...

sgrankin replied:
Good catch!

sgrankin commented:
Updated with CR fixes and rebased on latest master.  
Note that io.buoyant.grpc.interop.NetworkedInteropTest seems to be flaky (failed, then passed on re-run).

adleong commented on k8s/src/main/scala/io/buoyant/k8s/Watchable.scala:
```diff
@@ -83,19 +84,19 @@ private[k8s] abstract class Watchable[O <: KubeObject: TypeReference, W <: Watch
     fieldSelector: Option[String] = None,
     resourceVersion: Option[String] = None
-  ): (AsyncStream[W], Closable) = {
+  ): (Activity.State[W] => Unit) => Closable = { (state: Activity.State[W] => Unit) =>
```
Could this return type just be `Activity[W]` instead?  I think this should be as simple as moving `Activity(Var.async(...) { ... })` from the caller into this method.  That would make this method signature a little easier to understand.

sgrankin replied:
IIRC I actually started this change with watch returning a Event[].  It was indeed cleaner, and worked in deployment, but the tests failed:
- some of the existing tests write multiple events into the response to be processed before actually testing what came out on the other end.
- the only way to consume a stream of events from a Var or Activity seems to be Var.changes or Activity.states (which then calls Var.changes)
- in the observe method of the Var created in Var.async, I *think* multiple updates were managing to sneak in between the synchronized creation of the nested Var and the call to .observe which would register the Observer, so the first update would not necessarily be seen
- and so the tests would not necessarily see the first event and would fail.

I thought this behavior of missing some initial events should be ok in the wild (we'd get the latest event eventually), but wanted to make sure the tests (that test the `watch` function directly) saw the complete stream of events, as they did previously with AsyncStream.  I was also trying to avoid changing the logic of the tests.  So I ended up with this function  thus this function and the concurrent-queue machinery in the tests. 

I don't see a way of extracting the complete stream of events out of a Var (otherwise we'd likely run into the same issue as with AsyncStream, right?).  Anything I may be missing about Vars/Activities?

Actually... another option to improve the readability without changing the signature much (obvious in retrospect):
```
def watch(
    labelSelector: Option[String] = None,
    fieldSelector: Option[String] = None,
    resourceVersion: Option[String] = None,
    state: Activity.State[W] => Unit
  ) 
```
The watchable call-site doesn't actually benefit from being able to partially watch, and the ApiTest callsite is easily changed to work with this signature as well.

Thoughts?   It'd still keep the weird `state` type but make the `watch` signature plainer, and the tests can still avoid dealing with Var quirks.

sgrankin replied:
per discussion, went with the l updated signature

hawkw commented:
@sgrankin yeah, that's one of a handful of integration tests that are known to be somewhat flaky on CI (https://github.com/linkerd/linkerd/issues/1504), please feel free to restart any builds failed by that test!

adleong commented on k8s/src/main/scala/io/buoyant/k8s/Watchable.scala:
```diff
@@ -139,71 +121,106 @@ private[k8s] abstract class Watchable[O <: KubeObject: TypeReference, W <: Watch
+    def _processEventStream(
+      stream: () => AsyncStream[W],
+      resourceVersion: Option[String],
+      previousEvent: Option[W] = None
```
consider calling this "largestVersion" or something since it may not actually be the previous event

sgrankin replied:
updated to 'largestVersion' and 'largestEvent'.  

Aside: a good cleanup to do afterwards may be to get rid of the event argument here.  I think the resource version comparison from ResourceVersionOrdering can be factored out and used directly with the resource version here, avoiding the need to keep a whole event around.  (I think this may be the only use case of that Ordering type class ...  I didn't do the change wanting to limit the diff size.)

adleong commented on k8s/src/main/scala/io/buoyant/k8s/Watchable.scala:
```diff
@@ -139,71 +121,106 @@ private[k8s] abstract class Watchable[O <: KubeObject: TypeReference, W <: Watch
+          import Ordering.Implicits._
+          // Register the update only if its resource version is larger than the largest version
+          // seen so far.
+          if (previousEvent.forall((_: W) < event)) {
```
is the `_: W` type annotation necessary?

sgrankin replied:
It was in some other incarnation of this lambda, but it seems no longer (or was possibly an IntelliJ-only error).  Will remove.

hawkw commented on k8s/src/main/scala/io/buoyant/k8s/Watchable.scala:
```diff
@@ -139,71 +121,106 @@ private[k8s] abstract class Watchable[O <: KubeObject: TypeReference, W <: Watch
+        // Special case to handle Kubernetes bug where "too old resource version" errors are
+        // returned with status code 200 rather than status code 410.
+        // see https://github.com/kubernetes/kubernetes/issues/35068 for details
+        case Some((e: Watch.Error[O], _)) if e.status.code.contains(410) =>
```
Thanks for flattening the match arms like I asked, this looks great. The rephrased comments are much clearer too!

hawkw commented on k8s/src/test/scala/io/buoyant/k8s/v1/ApiTest.scala:
```diff
@@ -432,30 +431,77 @@ class ApiTest extends FunSuite
+      await(w.write(modified0))
+      poll { case None => }
+
+      // write an earlier event: no update since resource version is too low
```
Thanks for adding this test!

adleong commented on k8s/src/main/scala/io/buoyant/k8s/Watchable.scala:
```diff
@@ -139,71 +121,106 @@ private[k8s] abstract class Watchable[O <: KubeObject: TypeReference, W <: Watch
+     */
+    def _processEventStream(
+      stream: () => AsyncStream[W],
+      resourceVersion: Option[String],
```
I think this variable shadows the resourceVersion passed to `watch` which is a bit confusing.

adleong commented:
Ah, that's super interesting!  Yes, I think you are correct that it is a fundamental property of `Var` (and `Event`) that they do not keep history.  With that in mind, your approach here makes sense.  I do like the updated signature you propose (moving state into the param list).

siggy commented:
Hi @sgrankin, just checking in to see how things are going. We'd love to get this merged for our 1.3.3 release this week.

sgrankin commented:
I just got home :) 
I’ll do the last round of changes sometime in the next few hours — since these are all effectively cosmetic change stability should be unaffected (in case you are concerned about re-running the stress tests).

> On Nov 29, 2017, at 21:14, Andrew Seigner <notifications@github.com> wrote:
> 
> Hi @sgrankin <https://github.com/sgrankin>, just checking in to see how things are going. We'd love to get this merged for our 1.3.3 release this week.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/linkerd/linkerd/pull/1714#issuecomment-348060898>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AA9_Qef1LhzVyqNKALm8_lELx1Pmrippks5s7g-HgaJpZM4QrFLl>.
>

sgrankin commented:
@siggy: I'm done with all the changes. Thanks!
